File: backend/app/schemas/financials.py
from __future__ import annotations

import uuid
from decimal import Decimal
from pydantic import BaseModel


class IncomeStatementItem(BaseModel):
    period_year: int
    period_type: str
    period_quarter: int | None
    revenue_cr: Decimal | None
    ebitda_cr: Decimal | None
    ebitda_margin_pct: Decimal | None = None
    pat_cr: Decimal | None
    net_profit_margin_pct: Decimal | None = None
    eps_basic: Decimal | None
    eps_diluted: Decimal | None
    dividend_per_share: Decimal | None

    model_config = {"from_attributes": True}


class BalanceSheetItem(BaseModel):
    period_year: int
    period_type: str
    total_assets_cr: Decimal | None
    shareholders_equity_cr: Decimal | None
    long_term_debt_cr: Decimal | None
    short_term_debt_cr: Decimal | None
    cash_cr: Decimal | None
    receivables_cr: Decimal | None
    inventories_cr: Decimal | None

    model_config = {"from_attributes": True}


class CashFlowItem(BaseModel):
    period_year: int
    period_type: str
    cfo_cr: Decimal | None
    cfi_cr: Decimal | None
    cff_cr: Decimal | None
    capex_cr: Decimal | None
    free_cash_flow_cr: Decimal | None

    model_config = {"from_attributes": True}


class KeyRatioItem(BaseModel):
    period_year: int
    period_type: str
    roe_pct: Decimal | None
    roce_pct: Decimal | None
    roa_pct: Decimal | None
    ebitda_margin_pct: Decimal | None
    net_profit_margin_pct: Decimal | None
    pe_ratio: Decimal | None
    pb_ratio: Decimal | None
    ev_ebitda: Decimal | None
    debt_equity_ratio: Decimal | None
    current_ratio: Decimal | None
    interest_coverage: Decimal | None
    revenue_growth_pct: Decimal | None
    pat_growth_pct: Decimal | None
    cash_conversion_cycle: Decimal | None

    model_config = {"from_attributes": True}


class FinancialSummary(BaseModel):
    """All financial statements combined — used by the company financials page."""
    company_id: uuid.UUID
    income_statements: list[IncomeStatementItem]
    balance_sheets: list[BalanceSheetItem]
    cash_flows: list[CashFlowItem]
    key_ratios: list[KeyRatioItem]

File: frontend/src/types/financials.ts
export interface IncomeStatementItem {
  period_year: number;
  period_type: string;
  period_quarter: number | null;
  revenue_cr: number | null;
  ebitda_cr: number | null;
  ebitda_margin_pct: number | null;
  pat_cr: number | null;
  net_profit_margin_pct: number | null;
  eps_basic: number | null;
  eps_diluted: number | null;
  dividend_per_share: number | null;
}

export interface BalanceSheetItem {
  period_year: number;
  period_type: string;
  total_assets_cr: number | null;
  shareholders_equity_cr: number | null;
  long_term_debt_cr: number | null;
  short_term_debt_cr: number | null;
  cash_cr: number | null;
  receivables_cr: number | null;
  inventories_cr: number | null;
}

export interface CashFlowItem {
  period_year: number;
  period_type: string;
  cfo_cr: number | null;
  cfi_cr: number | null;
  cff_cr: number | null;
  capex_cr: number | null;
  free_cash_flow_cr: number | null;
}

export interface KeyRatioItem {
  period_year: number;
  period_type: string;
  roe_pct: number | null;
  roce_pct: number | null;
  roa_pct: number | null;
  ebitda_margin_pct: number | null;
  net_profit_margin_pct: number | null;
  pe_ratio: number | null;
  pb_ratio: number | null;
  ev_ebitda: number | null;
  debt_equity_ratio: number | null;
  current_ratio: number | null;
  interest_coverage: number | null;
  revenue_growth_pct: number | null;
  pat_growth_pct: number | null;
  cash_conversion_cycle: number | null;
}

export interface FinancialSummary {
  company_id: string;
  income_statements: IncomeStatementItem[];
  balance_sheets: BalanceSheetItem[];
  cash_flows: CashFlowItem[];
  key_ratios: KeyRatioItem[];
}
